### PR TITLE
Fix native x86 signed div and mod

### PR DESCRIPTION
**Pull request by Radeonares32**
Fix #XXX: Fix native x86 signed div and mod

## Summary

Fix native x86 division and modulo for negative dividends.

The native x86 backend uses `IDIV`, which expects the signed dividend in `EDX:EAX`. The previous lowering zeroed `EDX`, which only works for non-negative dividends and produces incorrect results for negative values.

## What has changed

- Added a `cdq()` x86 helper.
- Changed native x86 `div` and `mod` lowering to sign-extend `EAX` into `EDX:EAX` before `IDIV`.
- Added a custom native regression test for signed div/mod cases.

## Testing

- [ ] Added / updated tests under `tests/pos/` or `tests/warn/`
- [ ] Ran `./ci` locally and all tests pass
- [ ] Docs updated if the change affects user-visible behavior
- [x] All commits are signed off ([why?](https://github.com/typescope/jo/blob/main/CONTRIBUTING.md#contribution-terms))

Ran:

- `./build`
- `tests/custom/x86-signed-div-mod/test.sh`

The custom test checks:

- `bin/jo eval`
- `bin/jo compile --stack`
- `bin/jo compile --reg`

## Security impact

No direct security impact. This is a native x86 backend correctness fix for signed integer division and modulo.

---

## Implementation checklist (for new language features)

Not applicable; this PR does not add or change language features.

### CODE REVIEW

File: stack-lang/native/arch/X86.scala [hunk 1]
```diff
@@ -386,16 +386,15 @@ object X86 extends Assembler:
               move(Reg(rv), freeReg)
               freeReg
```
Comment (liufengyun) [inline stack-lang/native/arch/X86.scala:387]: Thank you for working on the TODO items @Radeonares32 👍

File: tests/custom/x86-signed-div-mod/app.jo [hunk 2]
```diff
@@ -0,0 +1,7 @@
+def main =
+  println (-7 / 2)
+  println (-7 % 2)
+  println (7 / -2)
+  println (7 % -2)
+  println (-7 / -2)
+  println (-7 % -2)
```
Comment (liufengyun) [inline tests/custom/x86-signed-div-mod/app.jo]: What is the motivation to use a custom test, instead of a normal `tests/pos`?

Comment (liufengyun) [reply tests/custom/x86-signed-div-mod/app.jo]: FYI, all tests under `tests/pos` are run by 6 backends: reference interpreter, JS, Ruby, Python, stack machine and register machine.

Comment (Radeonares32) [reply tests/custom/x86-signed-div-mod/app.jo]: Good question. I used a custom test because this PR is only fixing the native x86 lowering.

These negative div/mod cases also reveal an existing cross-backend difference: after this fix, `eval`, `--stack`, and `--reg` agree, but JS/Python/Ruby still differ from the interpreter for some negative `/` and `%` cases. Putting this directly under `tests/pos` would make those backends fail too.

So I kept the regression native-focused for this PR.

Comment (liufengyun) [reply tests/custom/x86-signed-div-mod/app.jo]: > JS/Python/Ruby still differ from the interpreter for some negative / and % cases

Indeed, languages differ here:

- Java: `-7 / 2 = -3`
- Python: `-7 // 2 = -4`
- JS: `Math.floor(-7 / 2) = -4` but `(-7 / 2) | 0 = -3`
- Ruby: `-7 / 2 = -4`

My understanding of the phenomena is that such difference does not matter in practice --- it's impossible to point fingers to say which language is wrong, and it's rarely used in practice (floating point is preferred for division when negative numbers involved).

Also note that the `Int` is platform integer (JS only has number), they are allowed to diff reasonably on edge cases related to overflow.

So for the current test, if we adapt it to test `-8 / 2`, then all backends should agree here.

Comment (Radeonares32) [reply tests/custom/x86-signed-div-mod/app.jo]: That makes sense for keeping this PR focused. I can adapt the regression to cases where all current backends agree.

That said, I do think the broader cross-backend difference is worth tracking separately. For a general-purpose language this might be just an edge case, but Jo’s security story seems to rely on code being constrained and predictable across execution targets. If the same source program can make different integer decisions depending on the backend, that could become surprising in policy-like code, quota checks, indexing, hashing, or capability-adjacent logic.

I agree this does not need to block this PR, and I am happy to keep this change focused on the native x86 `IDIV` issue. I just think it would be safer to record the cross-backend division/modulo semantics as a follow-up rather than treating it as irrelevant.

Comment (liufengyun) [reply tests/custom/x86-signed-div-mod/app.jo]: > I just think it would be safer to record the cross-backend division/modulo semantics as a follow-up rather than treating it as irrelevant.

Yes, it makes sense to track them as a separate issue. For now, the officially supported Ruby/Python agree when negative numbers are involved in division/modulo.

Comment (Radeonares32) [reply tests/custom/x86-signed-div-mod/app.jo]: Got it, that distinction helps. I was treating `eval` as the reference behavior when comparing backend outputs, but if Ruby/Python are the officially supported behavior here, then it makes sense not to pull that into this PR.

I will keep this PR focused on the native x86 issue and adjust the regression to cases where the supported backends agree.

Comment (liufengyun) [reply tests/custom/x86-signed-div-mod/app.jo]: Yes, we could in a separate PR to change the reference interpreter to use `Math.floorDiv` and `Math.floorMod`, and optionally change Native semantics too to floor semantics.

**Review by liufengyun — APPROVED** [review]
LGTM, thank you @Radeonares32 💯
